### PR TITLE
Use sqlalchemy ORM instead of string queries

### LINKED ISSUE #67: Use sqlalchemy instead of psycopg2 and Oracle_Cx

### PR DESCRIPTION
**Pull request by Jasopaum**
Closes #67

### CODE REVIEW

File: test/unit/analyze/test_mapping.py [hunk 1]
```diff
@@ -192,38 +196,43 @@ def test_find_cols_joins_maps_scripts(patient_mapping, fhir_concept_map_identifi
...
-        "patients.dod",
+        SqlColumn("patients", "row_id"),
+        SqlColumn("patients", "gender"),
+        SqlColumn("patients", "dob"),
+        SqlColumn("patients", "dod"),
+        SqlColumn("admissions", "marital_status"),
+        SqlColumn("patients", "expire_flag"),
+    }
+    for j in joins:
+        print(j.column1, j.column2)
```
Comment (simonvadee) [inline test/unit/analyze/test_mapping.py]: remove

File: fhirpipe/analyze/analyzer.py [hunk 2]
```diff
@@ -11,7 +11,9 @@ def analyze(self, resource_mapping):
         analysis = Analysis()
 
         # Get primary key table
-        analysis.primary_key_table, analysis.primary_key_column = get_primary_key(resource_mapping)
+        analysis.primary_key_table, analysis.primary_key_column = get_primary_key(
```
[the remark this answers is not in the document]
Comment (Jasopaum) [reply fhirpipe/analyze/analyzer.py]: I had this, plus the pk table which was a string but it's not needed anymore, I removed it

File: fhirpipe/extract/extractor.py [hunk 3]
```diff
@@ -18,22 +29,66 @@ def __init__(self, credentials, chunksize=None):
...
+            query = query.filter(self.get_column(pk_column).in_(pk_values))
+
+        return query
+
+    def get_columns(self, columns):
+        return [self.get_column(col) for col in columns]
+
+    def get_column(self, column):
+        table = self.get_table(column.table, column.owner)
+        return table.c.__getattr__(column.column)
```
[the remark this answers is not in the document]
Comment (Jasopaum) [reply fhirpipe/extract/extractor.py]: Not a method, just table.c[column.column]. Guess I didn't try hard enough

File: fhirpipe/extract/extractor.py [hunk 4]
```diff
@@ -18,22 +29,66 @@ def __init__(self, credentials, chunksize=None):
         db_string = build_db_url(credentials)
 
         self.engine = create_engine(db_string)
+        self.metadata = MetaData(bind=self.engine)
+        self.session = sessionmaker(self.engine)()
```
[the remark this answers is not in the document]
Comment (Jasopaum) [reply fhirpipe/extract/extractor.py]: SQLAlchemy stuff...
self.metadata is needed to have the right attributes for classes (ie the class for table Patient should have attributes subject_id, dob, etc.)
self. session is needed to do ORM queries

File: fhirpipe/extract/extractor.py [hunk 5]
```diff
@@ -18,22 +29,66 @@ def __init__(self, credentials, chunksize=None):
...
+        alchemy_cols = self.get_columns(columns)
+        base_query = self.session.query(*alchemy_cols)
+        query_w_joins = self.apply_joins(base_query, joins)
+        query_w_filters = self.apply_filters(query_w_joins, resource_mapping, pk_column, pk_values)
+
+        return query_w_filters.statement
+
+    def apply_joins(self, query, joins):
+        for join in joins:
+            foreign_table = self.get_table(join.column2.table, join.column2.owner)
```
[the remark this answers is not in the document]
Comment (Jasopaum) [reply fhirpipe/extract/extractor.py]: Don't think so, we're getting a sqlalchemy table here

File: fhirpipe/extract/extractor.py [hunk 6]
```diff
@@ -18,22 +29,66 @@ def __init__(self, credentials, chunksize=None):
...
+
+    def apply_filters(self, query, resource_mapping, pk_column, pk_values):
+        if resource_mapping["filters"]:
+            for filter in resource_mapping["filters"]:
+                col_name = filter["sqlColumn"]
+                col = self.get_column(
+                    SqlColumn(col_name["table"], col_name["column"], col_name["owner"])
+                )
+                rel_method = SQL_RELATIONS_TO_METHOD[filter['relation']]
+                query = query.filter(col.__getattr__(rel_method)(filter['value']))
```
Comment (achauve) [inline fhirpipe/extract/extractor.py]: rather use `getattr(col, rel_method)`

File: fhirpipe/analyze/sql_column.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    def __str__(self) -> str:
+        if self.owner:
+            return f"{self.owner}.{self.table}.{self.column}"
+        else:
+            return f"{self.table}.{self.column}"
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def table_name(self) -> str:
```
[the remark this answers is not in the document]
Comment (Jasopaum) [reply fhirpipe/analyze/sql_column.py]: I still do, unfortunately... I'll try to remove it in one of the next refactor

File: fhirpipe/extract/extractor.py [hunk 8]
```diff
@@ -18,22 +29,66 @@ def __init__(self, credentials, chunksize=None):
...
+        logging.info(f"sql query: {query}")
 
-        # Build the sql query
-        sql_query = build_sql_query(
-            analysis.cols, analysis.joins, analysis.primary_key_table, sql_filters
-        )
-        logging.info("sql query: %s", sql_query)
+        return run_sql_query(self.engine, query)
+
+    def sqlalchemy_query(self, columns, joins, table_name, pk_column, resource_mapping, pk_values):
```
[the remark this answers is not in the document]
Comment (Jasopaum) [reply fhirpipe/extract/extractor.py]: For the sake of separation of concerns

File: fhirpipe/extract/extractor.py [hunk 9]
```diff
@@ -18,22 +29,66 @@ def __init__(self, credentials, chunksize=None):
...
+        query_w_filters = self.apply_filters(query_w_joins, resource_mapping, pk_column, pk_values)
+
+        return query_w_filters.statement
+
+    def apply_joins(self, query, joins):
+        for join in joins:
+            foreign_table = self.get_table(join.column2.table, join.column2.owner)
+            query = query.join(
+                foreign_table,
+                self.get_column(join.column2) == self.get_column(join.column1),
```
[the remark this answers is not in the document]
Comment (Jasopaum) [reply fhirpipe/extract/extractor.py]: That's how we write a join with sqlalchemy.
I want to join foreign_table on join.left == join.right

**Pull request merged**